### PR TITLE
Prepare cache to change it by configuration

### PR DESCRIPTION
**Pull request by cbickel**
This PR changes the Cache to be a class instead of a trait.
With this change, it will be easier to change to another cache by configuration.

The configuration or the implementation of another cache is not part of this PR.

PG3#892 is 🔵 
PG4#593 is 🔵

### CODE REVIEW

File: common/scala/src/main/scala/whisk/core/database/MultipleReadersSingleWriterCache.scala [hunk 1]
```diff
@@ -146,138 +142,127 @@ trait MultipleReadersSingleWriterCache[W, Winfo] {
...
+
+            actualEntry.state.get match {
+                case Cached =>
+                    logger.debug(this, "cached read")
+                    makeNoteOfCacheHit(key)
+                    actualEntry.unpack.map { entry =>
+                        if (entry.getClass != mw.runtimeClass) {
+                            throw DocumentTypeMismatchException(s"document type ${entry.getClass} did not match expected type ${mw.runtimeClass}.")
                         }
+                        entry.asInstanceOf[Wsub]
```
Comment (markusthoemmes) [inline common/scala/src/main/scala/whisk/core/database/MultipleReadersSingleWriterCache.scala]: Is this needed as a runtime check? Can this be checked statically?

Comment (cbickel) [inline common/scala/src/main/scala/whisk/core/database/MultipleReadersSingleWriterCache.scala]: As the Cache is a cache for all WhiskEntities, there are Rules, Triggers and Actions in one Cache.
On getting an element by id, we should be sure, that the element the cache found has the expected type.

Comment (cbickel) [inline common/scala/src/main/scala/whisk/core/database/MultipleReadersSingleWriterCache.scala]: It's the same like getting the WhiskEntity out of the Datastore:
https://github.com/apache/incubator-openwhisk/blob/master/common/scala/src/main/scala/whisk/core/database/CouchDbRestStore.scala#L148

File: common/scala/src/main/scala/whisk/core/database/CouchDbStoreProvider.scala [hunk 2]
```diff
@@ -23,20 +23,25 @@ import whisk.common.Logging
...
         require(config != null && config.isValid, "config is undefined or not valid")
         require(config.dbProvider == "Cloudant" || config.dbProvider == "CouchDB", "Unsupported db.provider: " + config.dbProvider)
         assume(Set(config.dbProtocol, config.dbHost, config.dbPort, config.dbUsername, config.dbPassword, name(config)).forall(_.nonEmpty), "At least one expected property is missing")
 
-        new CouchDbRestStore[D](config.dbProtocol, config.dbHost, config.dbPort.toInt, config.dbUsername, config.dbPassword, name(config))
+        new CouchDbRestStore[D, CacheAbstraction](config.dbProtocol, config.dbHost, config.dbPort.toInt, config.dbUsername, config.dbPassword, name(config), cache)
+    }
+
+    def makeCache[CacheAbstraction](): WhiskCache[CacheAbstraction, DocInfo] = {
+        new MultipleReadersSingleWriterCache[CacheAbstraction, DocInfo]()
```
Comment (markusthoemmes) [inline common/scala/src/main/scala/whisk/core/database/CouchDbStoreProvider.scala]: Should we move this implementation to `ArtifactStoreProvider`? After all, it's not relying on implementation details.

**Comment by cbickel** [comment]
Replaced by #2624

**Pull request closed without merging**